[Sherwin Poh Kai Xun] iP

sherrpass opened the pull request:
### DukeMaster

> “You miss 100% of the shots you don’t take.” - [Kobe Bryant](https://en.wikipedia.org/wiki/Kobe_Bryant)

DukePro frees your mind of having to remember things you need to do. It's,

- Amazing
- Fast
- Convenient

All you need to do is to is,

1. Download it right [here](https://github.com/sherrpass/ip/releases/tag/A-Jar)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is FREE!

Features:

- [x]  Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)
- [ ]  Be actually useful (coming I hope)

Here's the main method that sparks the program off!

```
public static void main(String[] args) {
    new Duke("./data", "./data/tasks.txt").run();
}
```

Commit:
Add Gradle support

sherrpass committed:
Implemented skeletal functionality to Duke, echos all, exits with bye

sherrpass committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested

sherrpass committed:
Add the ability to mark tasks as done.

sherrpass committed:
Add support for tracking three types of tasks: Todo, Event and Deadline

sherrpass committed:
semi-automate the testing of Duke

sherrpass committed:
Add errors such as incorrect inputs entered by the user

sherrpass committed:
Add Delete functionality

sherrpass committed:
Tidy up code

sherrpass committed:
Add Save to and Load from capabilities from hard disk

sherrpass committed:
Add dates for deadlines and events, add new occuring function

sherrpass committed:
Merge branch 'branch-Level-7'

sherrpass committed:
Merge branch 'branch-Level-8'

sherrpass committed:
Extract out closely related code as classes

Ui: deals with interactions with the user
Storage: deals with loading tasks from the
file and saving tasks in the file
Parser: deals with making sense of the user
command
TaskList: contains the task list e.g., it
has operations to add/delete tasks in the list
Command: contains all commands

sherrpass committed:
Shift all files into Duke package

sherrpass committed:
Add Junit Test for Storage and Deadline classes

sherrpass committed:
Add Javadoc

sherrpass committed:
Fix styling and naming to fit coding standards

sherrpass committed:
Add functionality to find tasks by queries.

sherrpass committed:
Merge branch 'branch-A-JavaDoc'

sherrpass committed:
Merge branch 'branch-A-CodingStandard'

sherrpass committed:
Merge branch 'branch-Level-9'

kevinchua6 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,99 @@
+     * @param data_folder_path Path to the data folder.
+     * @param data_file_path Path to the data file.
+     */
+    public Storage(String data_folder_path, String data_file_path) {
```
Should variables be in camelCase?

kevinchua6 commented on src/test/java/duke/StorageTest.java:
```diff
@@ -0,0 +1,71 @@
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class StorageTest {
+    private final static String TEST_DATA_FOLDER_PATH = "./dataTest";
```
I think you shouldn't use `private final static`, but `private static final`: https://google.github.io/styleguide/javaguide.html

kevinchua6 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,50 @@
+        return (isDone ? "X" : " "); // mark done task with X
+    }
+
+    public boolean getDoneStatus() {
```
Shouldn't boolean methods sound like booleans? Any reason why you didn't do something like `isDone` or `hasCompleted`?

kevinchua6 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,50 @@
+        return "|" + getStatusIcon() + "|" + description;
+    }
+
+    public boolean matches(String query) {
```
Again, shouldn't boolean methods should sound like booleans? Perhaps a more intuitive or explanatory method name here?

kevinchua6 commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,7 @@
+package duke;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Should we use individual imports instead of wildcard imports?

kevinchua6 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,91 @@
+        ArrayList<Task> matchingTasks = tasks.getMatchingTasks(query);
+        showTasks(matchingTasks);
+    }
+}
```
I like that you have a consistent naming system for methods!

kevinchua6 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,49 @@
+        matchingTasks.removeIf(task -> !task.matches(query));
+        return matchingTasks;
+    }
+}
```
I like that you use verbs for all your method names!

kevinchua6 reviewed (commented):
Good job overall!

wlren commented on src/main/java/duke/AddCommand.java:
```diff
@@ -0,0 +1,59 @@
+    public void execute(TaskList tasks, Storage storage, Ui ui) throws DukeException {
+        Task newTask;
+        switch (taskType) {
+            case "todo":
+                newTask = new Todo(description);
+                break;
+            case "deadline":
+                newTask = new Deadline(description, date);
+                break;
+            case "event":
+                newTask = new Event(description, date);
+                break;
+            default:
+                throw new DukeException("Invalid task type provided!");
```
Coding standard for CS2103 requires switch statements to not have tab for case, take note to edit the default IDE settings!

wlren commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,58 @@
+ * Represents a deadline task.
+ */
+public class Deadline extends Task {
+    protected LocalDate by;
```
Should perhaps be private since there is already a getter method that exposes this variable

wlren commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,58 @@
+        return "D" + super.toStringData() + "|" + by;
+    }
+
+    @Override
+    public boolean matches(String query) {
+        return super.matches(query)
+                || by.format(DateTimeFormatter.ofPattern("MMM dd yyyy")).toLowerCase().contains(query.toLowerCase())
+                || query.equalsIgnoreCase("deadline");
+    }
```
A comment to explain this logic would be great

wlren commented on src/main/java/duke/DeleteCommand.java:
```diff
@@ -0,0 +1,34 @@
+     * @param ui UI component.
+     */
+    @Override
+    public void execute(TaskList tasks, Storage storage, Ui ui) {
```
```suggestion
    public void execute(TaskList tasks, Storage storage, Ui ui) throws DukeException {
```

wlren commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,58 @@
+ * The Duke Application.
+ */
+public class Duke {
+    private Storage storage;
+    private TaskList tasks;
+    private Ui ui;
```
Can be private final

wlren commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,68 @@
+                throw new DukeException("I'm sorry, but I don't know what that means :-(");
+            }
+        } catch (DateTimeParseException e) {
+            throw new DukeException("Date provided should be in YYYY-MM-DD format.");
```
Is there perhaps a better way to do the error handling?

wlren reviewed (commented):
LGTM! Looks very clean and has good quality code overall, besides the comments that the previous reviewer gave, I identified a few small details to take note of, good job!

sherrpass committed:
Merge remote-tracking branch 'origin/add-gradle-support' into A-Gradle

sherrpass committed:
Add gradle

sherrpass committed:
Merge branch 'A-Gradle'

sherrpass committed:
Add Checkstyle to detect style violations, clean up violations

sherrpass committed:
Add GUI using JavaFC

sherrpass committed:
Merge branch 'Level-10'

sherrpass committed:
Add Assert feature

Current Duke codebase has many implicit assumptions
about the Internal Invariants and Control-Flow Invariants.

Unverified assumptions may cause bugs to be undiscovered
during development and cause issues for the user
in production.

Let's add assertions throughout the codebase to verify
these assumptions.

Thorough checking of assumptions using assertions can
ensure that bugs in the code are discovered quickly.

sherrpass committed:
Improve code quality

Current code has areas of improvement with regards
to readability and code duplication.

Poor readability would make the code less
maintainable and comprehensable to others. Code
duplication may lead to greater effort to
adjust duplicated code in future.

Implemenented SLAP, abstracted more logic to make
code at each level simpler, removed duplicated
code.

SLAP, abstractions and DRY principle makes code
more readable and improves code quality.

sherrpass committed:
Merge pull request #2 from sherrpass/branch-A-Assertions

Add Assert feature

sherrpass committed:
Merge branch 'master' into branch-A-CodeQuality

sherrpass committed:
Merge pull request #3 from sherrpass/branch-A-CodeQuality

Improve code quality

sherrpass committed:
Implement undo command

sherrpass committed:
Merge pull request #4 from sherrpass/branch-BCD-Extension

Implement undo command

sherrpass committed:
Support undoing multiple commands.

sherrpass committed:
Add extra exception handling, change profile pictures, add ui screenshot

sherrpass committed:
Set theme jekyll-theme-slate

sherrpass committed:
Update user guide in README.md and renamed Duke class to DukeMaster

sherrpass committed:
Merge branch 'master' of https://github.com/sherrpass/ip

sherrpass committed:
Fix header issue in user guide

sherrpass committed:
Add response message to list command if task list is empty

sherrpass committed:
Add new DukeDataException subclass and error handling for invalid data file

sherrpass committed:
Add undo to user guide.

sherrpass committed:
Clean up user guide.
